libarchive/archive_read_support_format_cpio.c

```diff
@@ -633,6 +633,13 @@ header_newc(struct archive_read *a, struct cpio *cpio,
 	/* Pad name to 2 more than a multiple of 4. */
 	*name_pad = (2 - *namelength) & 3;
 
+	/* Make sure that the padded name length fits into size_t. */
+	if ((size_t)(*namelength + *name_pad) < *namelength) {
```

Comment:
The C standard defines unsigned arithmetic as being modulo 2^N:

"A computation involving unsigned operands can never overflow,
because a result that cannot be represented by the resulting unsigned integer type is
reduced modulo the number that is one greater than the largest value that can be
represented by the resulting type."

I'm fine with with your proposed solution, but I find "if (x + y < x)" to be a far more idiomatic way of expressing "does this sum get reduced modulo 2^N".